opencti-platform/Dockerfile

```diff
@@ -1,47 +1,53 @@
-FROM node:12.21.0-alpine
+FROM node:12.21.0-alpine3.12 AS base
 
-# Copy work files
-COPY opencti-front /opt/opencti-build/opencti-front
+
+FROM base AS graphql-deps-builder
+
+WORKDIR /opt/opencti-build/opencti-graphql
+COPY opencti-graphql/package.json opencti-graphql/yarn.lock ./
+RUN yarn install --frozen-lockfile --production && yarn cache clean --all
+
+
+FROM base AS graphql-builder
+
+WORKDIR /opt/opencti-build/opencti-graphql
+COPY opencti-graphql/package.json opencti-graphql/yarn.lock ./
+RUN yarn install --frozen-lockfile
 COPY opencti-graphql /opt/opencti-build/opencti-graphql
-COPY entrypoint.sh /
+RUN yarn run webpack --mode production
+
+
+FROM base AS front-builder
+
+WORKDIR /opt/opencti-build/opencti-front
+COPY opencti-front/package.json opencti-front/yarn.lock ./
+RUN yarn install --frozen-lockfile
+COPY opencti-front /opt/opencti-build/opencti-front
+COPY opencti-graphql/config/schema/opencti.graphql /opt/opencti-build/opencti-graphql/config/schema/opencti.graphql
+RUN yarn relay && yarn generate
+
+
+FROM base AS app
 
-# This hack is widely applied to avoid python printing issues in docker containers.
-# See: https://github.com/Docker-Hub-frolvlad/docker-alpine-python3/pull/13
+RUN set -ex; \
+    apk add --no-cache tini~=0.19 gcc~=9.3 musl-dev~=1.1 python3~=3.8 python3-dev~=3.8; \
+    python3 -m ensurepip; \
+    rm -rv /usr/lib/python*/ensurepip; \
+    pip3 install --no-cache-dir --upgrade pip setuptools wheel; \
+    ln -sf python3 /usr/bin/python;
+WORKDIR /opt/opencti
+COPY opencti-graphql/src/python/requirements.txt ./src/python/requirements.txt
+RUN pip3 install --no-cache-dir --requirement ./src/python/requirements.txt
+COPY --from=graphql-deps-builder /opt/opencti-build/opencti-graphql/node_modules ./node_modules
+COPY --from=graphql-builder /opt/opencti-build/opencti-graphql/build ./build
+COPY --from=front-builder /opt/opencti-build/opencti-front/build ./public
+COPY opencti-graphql/src ./src
+COPY opencti-graphql/config ./config
+COPY opencti-graphql/script ./script
 ENV PYTHONUNBUFFERED=1
+ARG MAX_OLD_SPACE_SIZE=1024
+ENV NODE_OPTIONS=--max_old_space_size=${MAX_OLD_SPACE_SIZE}
```

Comment:
You're right having a default max old space size specified in the image can be interesting. But it should be possible to override it in the docker compose. I'm not a docker specialist, does it still possible to override this var with this kind of DockerFile build?

The API today can run with 100Mo of memory but with the current usage of most of our customers, we find safer to setup 8192 as default. Its really hard to have any recommendation because like always it depends of too much factor.

Reply:
> does it still possible to override this var with this kind of DockerFile build?

Yes, it is always possible to override `ENV` specified in the dockerfile at runtime with any container orchestration.

> we find safer to setup 8192 as default

Alright, so should I go with 1, which can be overridden both at build time with docker build-args or at runtime with any orchestration passing the env `NODE_OPTIONS` to the container:
```Dockerfile
ARG MAX_OLD_SPACE_SIZE=8192
ENV NODE_OPTIONS=--max_old_space_size=${MAX_OLD_SPACE_SIZE}
```
or 2, specified straightforwardly which is only possible to overridden be at runtime with any orchestration passing the env `NODE_OPTIONS` to the container:
```Dockerfile
ENV NODE_OPTIONS=--max_old_space_size=8192
```